app/src/main/java/unipd/se18/ocrcamera/PreProcessing.java

```diff
@@ -0,0 +1,394 @@
+        Core.add(tempLuminance , blueLuminance , totalLuminance); //Temp + Blue = Luminance
+
+        //Calculate the sum of the values of all pixels
+        Scalar sum = Core.sumElems(totalLuminance);
+
+        //Determines the image's bit
+        int bit;
+        switch ( brightnessMat.depth() ) {
+            case CV_8U:  bit = 8; break;
+            case CV_8S:  bit = 8; break;
+            case CV_16U: bit = 16; break;
+            case CV_16S: bit = 16; break;
+            case CV_32S: bit = 32; break;
+            case CV_32F: bit = 32; break;
+            case CV_64F: bit = 64; break;
+            default: return BrightnessValue.IMAGE_IS_OK;
+        }
+        //Calculate the percentage of the brightness
+        double brightness = sum.val[0]/((Math.pow(2,bit)-1)*brightnessMat.rows()
+                *brightnessMat.cols())*2;
+
+        Log.d(TAG, "Brightness:"+brightness);
+
+        /*Bounds to define if the image is dark or bright.
+          The values were decided on the basis of various tests*/
+        double upperBound = 0.9;
+        double lowerBound = 0.4;
+
+        if (brightness > upperBound){             //Image is too bright
+            return BrightnessValue.IMAGE_TOO_BRIGHT;
+        } else if (brightness < lowerBound){        //Image is too dark
+            return BrightnessValue.IMAGE_TOO_DARK;
+        } else {      //Image is neither too bright nor too dark
+            return BrightnessValue.IMAGE_IS_OK;
+        }
+    }
+
+
+    /**
+     * Change the brightness of the image into an optimal one
+     * @param image the image we want to modify the brightness
+     * @return the image with the modified brightness
+     * @author Thomas Porro(g1), Giovanni Fasan(g1), Oscar Garrido(g1)
+     */
+    private Bitmap editBright(Bitmap image){
+        /*This variable is used to put a limit to the change of the image's brightness.
+          The value 240 is derived from the fact that in the for loop we try to modify
+          the value of all the pixels of a step, and being the maximum value = 255 (pixel's
+          color maximum value, we put the limit on 240*/
+        final int maxBrightness = 240;
+        final int step = 15;
+
+        //Converts the image into a matrix
+        Mat bright;
+        try{
+            bright = IPUtils.conversionBitmapToMat(image);
+        } catch (ConversionFailedException error){
+            Log.e(TAG, error.getErrorMessage());
+            return image;
+        }
+
+        //Call the internal method isBright to detect if the image is bright or dark
+        //and change the brightness according to the number obtained
+        Mat modifiedMat = new Mat();
+        switch (isBright(bright)) {
+            case IMAGE_TOO_BRIGHT:
+                Log.d(TAG, "Case==IMAGE_TOO_BRIGHT");
+                //Darkens the colour's brightness until it's in an optimal value
+                for(double changeBrightness = 0; changeBrightness != maxBrightness;
```

Comment:
THIS COULD CAUSE PROBLEMS! you should end the for if !=-maxBrightness (to be shoure i would put <=, somebody could decide to change step). As it is now it could go into an infinite loop